valgrind-cpp_stl

```diff
@@ -12,4 +12,36 @@ CPP_TEST_OUT_DIR="$2"
 
 OBJ_DIR="$SRC_DIR/bin"
 
-valgrind --leak-check=full --xml=yes --xml-file="$CPP_TEST_OUT_DIR/valgrind.xml" "$OBJ_DIR/ks_tests"
+run_valgrind() {
+	_attempt_no=$1
+	shift
+	_xml_file="$CPP_TEST_OUT_DIR/valgrind-${_attempt_no}.xml"
+	valgrind --leak-check=full --xml=yes --xml-file="${_xml_file}" "$OBJ_DIR/ks_tests" "$@"
+
+	# For some obscure reason, valgrind generates this file as 0600 (owner-readable only).
+	# This disrupts invocation from Docker pipeline, as after return back from the Docker container
+	# this will become root-owned and root-only-readable file. We fix it here.
+	chmod 644 "${_xml_file}"
+}
+
+read_next=1
+
+run_attempt=0
+set --
+
+while [ -n "${read_next}" ]; do
+	run_attempt=$((run_attempt+1))
+	run_valgrind "${run_attempt}" "$@" || :
+
+	# if the `aborted_tests-{N}.txt` file is empty, then there were no aborted
+	# tests and it's the last attempt we need to process
+	read_next=
+	while IFS= read -r line; do
+		read_next=1
+		set -- "$@" "--run_test=!${line}"
+	done < "${CPP_TEST_OUT_DIR}/aborted_tests-${run_attempt}.txt"
+done
```

Comment:
Again, one more argument towards re-doing the builds excluding offending file after run: we won't need  to maintain such complex magic around valgrind runs.

Reply:
We don't _need_ to do that now either - as far as the contents of `ci.json` is concerned at the moment, only the log from the last Valgrind run (excluding any aborted tests) is copied to `valgrind.xml` and [used by `aggregate/convert_to_json`](https://github.com/kaitai-io/kaitai_struct_tests/blob/ac7dd7f068a5ea8e3fb86b2ec9666ff12aeb5b76/aggregate/convert_to_json#L71). This is explained at the bottom of `valgrind-cpp_stl` ([`valgrind-cpp_stl:45-47`](https://github.com/kaitai-io/kaitai_struct_tests/blob/ac7dd7f068a5ea8e3fb86b2ec9666ff12aeb5b76/valgrind-cpp_stl#L45-L47)):

```sh
# Currently, aggregate/convert_to_json only processes "valgrind.xml", so we
# make a copy with that name of the last (successful) attempt's log
cp -v "$CPP_TEST_OUT_DIR/valgrind-${run_attempt}.xml" "$CPP_TEST_OUT_DIR/valgrind.xml"
```

However, we'll eventually want to aggregate all `valgrind-*.xml` logs that this new version of `valgrind-cpp_stl` produces to `ci.json`, because these Valgrind logs from running aborted tests contain all the details of the memory error causing the abortion, as I mentioned above in https://github.com/kaitai-io/kaitai_struct_tests/pull/105#discussion_r1536646230.

So although the generated `valgrind-*.xml` logs are not used at the moment, they can be already helpful when manually inspecting the CI artifacts, and it doesn't hurt to have them available.